rnnalyse/activations/activation_reader.py

```diff
@@ -22,34 +23,154 @@ class ActivationReader:
     Attributes
     ----------
     activations_dir : str
-    labels : np.ndarray
-        Numpy array containing the extracted labels
-    data_len : int
-        Number of extracted activations
+    label_path : str
+    activations : Optional[np.ndarray]
+        Numpy array of activations that are currently read into ram.
+    _labels : Optional[np.ndarray]
+        Numpy array containing the extracted labels. Accessed by the
+        property self.labels.
+    _data_len : int
+        Number of extracted activations. Accessed by the property
+        self.data_len.
+    _activation_ranges : Optional[ActivationRanges]
+        Dictionary mapping sentence keys to their respective location
+        in the .activations array.
     """
+
     def __init__(self,
                  activations_dir: str,
                  label_path: Optional[str] = None) -> None:
 
         self.activations_dir = trim(activations_dir)
 
-        self.labels = self._read_labels(label_path)
-        self.data_len = len(self.labels)
-
-    def _read_labels(self, label_path: Optional[str]) -> np.ndarray:
         if label_path is None:
             label_path = f'{self.activations_dir}/labels.pickle'
+        self.label_path = label_path
+
+        self._activations = None
+        self._labels: Optional[np.ndarray] = None
+        self._data_len: int = -1
+        self._activation_ranges: Optional[ActivationRanges] = None
+
+        self.activations = None
+
+    # TODO: Add examples here or somewhere else
+    def __getitem__(self, key: ActivationKey) -> np.ndarray:
+        """ Provides indexing of activations, indexed by sentence
+        position or key, or indexing the activations itself. Indexing
+        based on sentence returns all activations belonging to that
+        sentence.
+
+        Indexing by position ('pos') refers to the order of extraction,
+        selecting the first sentence ([0]) will return all activations
+        of the sentence that was extracted firstly.
 
-        return load_pickle(label_path)
+        Sentence keys refer to the keys of the labeled corpus that was
+        used in the Extractor.
+
+        Indexing can be done by key, index list/np.array, or slicing
+        (which is translated into a range of keys).
+
+        The index, indexing type and activation name can be provided as:
+          [index]
+        | [index, indextype]         | [index, a_name]
+        | [index, a_name, indextype] | [index, indextype, a_name]
+        With indextype either 'pos', 'key' or 'all', and activation name
+        a (layer, name) tuple.
+
+        If activationname is not provided it should have been set
+        beforehand like `reader.activations = activationname`.
```

Comment:
So if I understand this correctly, if `reader.activations` is set even only using `[index]` will return the activations corresponding to `activationname`?

Reply:
Exactly. My idea for this is that this allows you to simply set/read the activations only once, and have all your subsequent indexing being applied to those activations.